inc/class-comment-popularity.php

```diff
@@ -609,18 +652,42 @@ public function comment_vote( $user_id, $comment_id, $vote ) {
 
 		// see if user has already voted
 		$logged_votes = $this->get_visitor()->retrieve_logged_votes();
-		if ( is_array( $logged_votes ) && array_key_exists( 'comment_id_' . $comment_id, $logged_votes ) ) {
-			$last_action = $logged_votes[ 'comment_id_' . $comment_id ]['last_action'];
+		if ( is_array( $logged_votes ) && array_key_exists( $comment_id, $logged_votes ) ) {
+			$last_action = $logged_votes[ $comment_id ][HMN_CP_Visitor::LOGGED_VOTES_ACTION_KEY];
+
+			$vote_value = ( 'upvote' == $last_action ) ?
+				$this->get_vote_value( 'downvote' ) : $this->get_vote_value( 'upvote' );
+
+			if (
+				$vote == 'upvote' && $last_action == 'downvote'
+				|| $vote == 'downvote' && $last_action == 'upvote'
+			) {
+				$vote_value *= 2;
+				$vote_count = 0;
+			}
+		}
 
-			if ( 'undo' === $labels [ $vote ] ) {
-				// undo the previous action
-				$this->get_visitor()->unlog_vote( $comment_id, $last_action );
+		if ($vote == 'undo') {
+			$vote_count = -1;
+		}
 
-				$vote_value = ( 'upvote' === $last_action ) ? $this->get_vote_value( 'downvote' ) : $this->get_vote_value( 'upvote' );
-			}
+		if ( false === $this->update_comment_weight( $comment_id, $vote_value, $vote_count ) ) {
```

Comment:
We prefer to avoid calling functions inside conditional checks especially if they modify data, it's to easier to see and clearer what's happening to do this before the `if()` check.

Additionally you introduced `WP_Error` as a possible return value but that result isn't handled here.